### PR TITLE
[CDAP-19437] Exclude recent log folders during cleanup in LogCleaner service and fix flaky test

### PR DESCRIPTION
**Pull request by vsethi09**
### Change Description
* [CDAP-19437](https://cdap.atlassian.net/browse/CDAP-19437) Exclude recent log folders during cleanups in LogCleaner service and clean-up stale files
  * Recent empty folders are excluded on the basis of date format.
  * Changed this to use last modified date of the folders.
  * Follow up of https://github.com/cdapio/cdap/pull/14403#discussion_r927018681
* [CDAP-19482](https://cdap.atlassian.net/browse/CDAP-19482) CDAPLogAppenderTest#testCDAPLogAppenderSizeBasedRotation fails intermittently

### Verification
* Verified log files and folder cleanup in CDAP sandbox

### CODE REVIEW

File: cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java [hunk 1]
```diff
@@ -99,21 +94,16 @@ private void cleanupLogFolders() {
...
-      Set<String> excludeFolders = new HashSet<String>();
-      for (int day = 0; day < EXCLUDE_FOLDER_CLEANUP_DAYS; day++) {
-        long time = currentTime - day * MILLISECONDS_IN_DAY;
-        String date = LogFileManager.formatLogDirectoryName(time);
-        excludeFolders.add(date);
+      // Exclude folders which were last modified within retention period.
+      // Clean up stale empty folders inside logs folder.
+      long checkpointTimestamp = startTime - retentionDurationMs;
+      for (Location location : logsDirectoryLocation.list()) {
+        cleanupEmptyFolders(location, checkpointTimestamp);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java]: So the folder being deleted may not be empty, correct? If that's the case, please rename the method appropriately.

File: cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java [hunk 2]
```diff
@@ -130,16 +120,16 @@ private void cleanupLogFolders() {
...
     if (location == null) {
       return true;
     }
     if (!location.isDirectory()) {
       // Parent directory has files.
       return false;
     }
-    if (excludeFolders.contains(location.getName())) {
-      // No need to traverse excluded folders
+    if (location.lastModified() > checkpointTimestamp) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java]: What would be the last modified date of a directory Location?

File: cdap-watchdog/src/test/java/io/cdap/cdap/logging/clean/LogCleanerTest.java [hunk 3]
```diff
@@ -227,33 +231,41 @@ public void testLogFolderCleanup(int folderCleanupBatchSize,
...
     logCleaner.run();
 
     // Assert for the remaining folder count
-    Assert.assertEquals(2, logFolder.list().size());
+    Assert.assertEquals(3, logFolder.list().size());
+  }
+
+  private void createFolderWithDateFormat(Location logFolder, int dayOffset) throws IOException {
+    long time = System.currentTimeMillis();
+    String date = new SimpleDateFormat("yyyy-MM-dd").format(new Date(time - dayOffset * MILLIS_IN_DAY));
```
Comment (chtyim) [inline cdap-watchdog/src/test/java/io/cdap/cdap/logging/clean/LogCleanerTest.java]: Should use the `LogFileManager.formatLogDirectoryName`?

Comment (vsethi09) [reply cdap-watchdog/src/test/java/io/cdap/cdap/logging/clean/LogCleanerTest.java]: Done.

File: cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java (hunk 1 above)
Comment (vsethi09) [reply cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java]: Renamed to 'cleanupFoldersIfEmpty'

File: cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java (hunk 2 above)
Comment (vsethi09) [reply cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java]: Last modified date of a directory is the date at which the directory was created or any file within that directory was modified (or created / deleted).

File: cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java [hunk 4]
```diff
@@ -130,26 +120,25 @@ private void cleanupLogFolders() {
    *         folders are cleaned up, true otherwise.
    * @throws IOException
    */
-  private boolean cleanupEmptyFolders(@Nullable Location location, Set<String> excludeFolders) throws IOException {
+  private boolean cleanupFoldersIfEmpty(@Nullable Location location, long checkpointTimestamp) throws IOException {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java]: I think the `location` will never be `null`, is it?

**Review by chtyim — APPROVED** [review]

File: cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java (hunk 4 above)
Comment (vsethi09) [reply cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java]: Cannot be null. Updated the code.

**Pull request merged**